ambry-clustermap/src/main/java/com/github/ambry/clustermap/InstanceConfigToDataNodeConfigAdapter.java

```diff
@@ -33,82 +38,146 @@
   private final HelixManager helixManager;
-  private final ClusterMapConfig clusterMapConfig;
+  private final Converter converter;
+  private final String clusterName;
+  private volatile HelixAdmin helixAdmin = null;
 
   /**
    * @param helixManager the {@link HelixManager} to use as the source of truth for {@link InstanceConfig}s.
    * @param clusterMapConfig the {@link ClusterMapConfig} to use.
    */
   public InstanceConfigToDataNodeConfigAdapter(HelixManager helixManager, ClusterMapConfig clusterMapConfig) {
     this.helixManager = helixManager;
-    this.clusterMapConfig = clusterMapConfig;
+    this.converter = new Converter(clusterMapConfig);
+    clusterName = clusterMapConfig.clusterMapClusterName;
   }
 
   @Override
   public void addDataNodeConfigChangeListener(DataNodeConfigChangeListener listener) throws Exception {
     helixManager.addInstanceConfigChangeListener((InstanceConfigChangeListener) (instanceConfigs, context) -> {
       Iterable<DataNodeConfig> dataNodeConfigs =
-          () -> instanceConfigs.stream().map(this::convert).filter(Objects::nonNull).iterator();
+          () -> instanceConfigs.stream().map(converter::convert).filter(Objects::nonNull).iterator();
       listener.onDataNodeConfigChange(dataNodeConfigs);
     });
   }
 
-  /**
-   * @param instanceConfig the {@link InstanceConfig} to convert to a {@link DataNodeConfig} object.
-   * @return the {@link DataNodeConfig}, or {@code null} if the {@link InstanceConfig} provided has an unsupported schema
-   *         version.
-   */
-  private DataNodeConfig convert(InstanceConfig instanceConfig) {
-    return convert(instanceConfig, clusterMapConfig);
+  @Override
+  public boolean set(DataNodeConfig config) {
+    InstanceConfig instanceConfig = converter.convert(config);
+    return getHelixAdmin().setInstanceConfig(clusterName, instanceConfig.getInstanceName(), instanceConfig);
   }
 
-  /**
-   * Exposed for testing.
-   * @param instanceConfig the {@link InstanceConfig} to convert to a {@link DataNodeConfig} object.
-   * @param clusterMapConfig the {@link ClusterMapConfig} containing any default values that may be needed.
-   * @return the {@link DataNodeConfig}, or {@code null} if the {@link InstanceConfig} provided has an unsupported schema
-   *         version.
-   */
-  static DataNodeConfig convert(InstanceConfig instanceConfig, ClusterMapConfig clusterMapConfig) {
-    int schemaVersion = getSchemaVersion(instanceConfig);
-    if (schemaVersion != 0) {
-      LOGGER.warn("Unknown InstanceConfig schema version {} in {}. Ignoring.", schemaVersion, instanceConfig);
-      return null;
+  @Override
+  public DataNodeConfig get(String instanceName) {
+    InstanceConfig instanceConfig = getHelixAdmin().getInstanceConfig(clusterName, instanceName);
+    return instanceConfig != null ? converter.convert(instanceConfig) : null;
+  }
+
+  void setHelixAdmin(HelixAdmin helixAdmin) {
+    this.helixAdmin = helixAdmin;
```

Comment:
Yes, this will be used in the next PR, so I will keep it. I was originally working on these two PRs on the same branch so they weren't fully isolated.